debug: implement `treeRepr()` for `PNode` (#191)
nim-works/nimskull: compiler/utils/tree_repr_doc.rst

Generate tree representation from `PNode`.

Generated output packs in a lot of information - here is a short list of
features, those will be elaborated on

- subnode index in relation to the parent node, optionally extended index
  (like`[0][0]`)
- symbol symbol kind, flags, other relevant elements
- type, if exists
- comment message (optionally)
- all node flags
- all IDs when applicable
- maximum target node depth and maximum node length (to avoid drowning in
  dumps when printing large entries)
- pack things as close as possible horizontally, or make them more spaced
  out vertically
- node declaration information
- any associated structured reports (IDs, kinds, generation location)
- symbol definition location if any
- type symbols definition location if any

Procedure arguments:

- `conf`: some configuration options are inferred from it, such as
  `--filenames:canonical`, but it is mostly used to get full paths from
  file ids
- `pnode`: node to be printed - can be `nil`, recursive, `nkError` or any
  other kind form or shape.
- `flags`: Set of the formatting options for the procedure. Several

>   built-in constants are defined in this module that migh tbe suitable for

Reviewer: ```suggestion
  built-in constants are defined in this module that might be suitable for
```

  different debugging or AST exploration tasks.
- `maxDepth`: Ignore all nodes that are placed deeper than that. Useful to

>   view only high-level overvew for large nodes, such as full proc bodies

Reviewer: ```suggestion
  see a high-level overview for large nodes, such as full proc bodies
```

> - `maxLength`: on each level, print only that much subnodes. Useful to cut

Reviewer: ```suggestion
- `maxLength`: on each level, print upto that many subnodes. Useful to cut
```

  out parts on large `case` trees, toplevel nodes for modules, statement
  lists and such.

.. note:: Output examples in the documentation might be slightly different

>           from the actual output due to

Reviewer: Missing next part of the sentence?

.. code-block:: literal

    Call
    0 Sym vmTarget (file.vmTarget)
        sk:   Proc
        flags:{Used, Global}
        typ:  Proc (arg): 
    1 ObjConstr
        nflags:{Sem}
        typ:  Object sk:Type
    1.0 Sym Obj (file.Obj)
          sk:   Type
          flags:{Used, Global}
          typ:  Object sk:Type

This is a tree representation listing from the end of the
`semexprs.semExpr` for this piece of code:

.. code-block:: nim

  vmTarget(Obj())

- `.0`, `.1` and `.1.0` show subnode positions of the analyzed input. This
  is very useful when you need to index into input tree - you can
  immediately see necessary positions.

  **TODO**: Allow printing out special defines, like `ast.pragmasPos`
  instead of the hardcoded indices

> - After node position somes the node symbol - `Call`, `Sym`, `ObjConstr`

Reviewer: Sorry, could figure out what you meant by "somes" to offer a suggestion.

Author: `comes`, but it took me another two mintes to figure this out as well
